// naive-ui/lib/tree/src/TreeNodeContent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vue_1 = require("vue");
exports.default = vue_1.defineComponent({
    name: 'TreeNodeContent',
    props: {
        clsPrefix: {
            type: String,
            required: true
        },
        disabled: {
            type: Boolean,
            default: false
        },
        onClick: Function,
        onDragstart: Function
    },
    setup(props) {
        const selfRef = vue_1.ref(null);
        function doClick(e) {
            const { onClick } = props;
            if (onClick)
                onClick(e);
        }
        function handleClick(e) {
            doClick(e);
        }
        return {
            selfRef,
            handleClick
        };
    },
    render() {
        const { clsPrefix, handleClick, onDragstart } = this;
        return (vue_1.h("span", { ref: "selfRef", class: [`${clsPrefix}-tree-node-content`], onClick: handleClick, draggable: onDragstart === undefined ? undefined : true, onDragstart: onDragstart },
            vue_1.h("div", { class: `${clsPrefix}-tree-node-content__text` }, this.$slots)));
    }
});
