// naive-ui/lib/_deprecated/service-layout/src/ServiceLayout.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vue_1 = require("vue");
const vooks_1 = require("vooks");
const vue_router_1 = require("vue-router");
const SiderHeader_1 = require("./SiderHeader");
const menu_1 = require("../../../menu");
const Layout_1 = require("../../../layout/src/Layout");
const LayoutSider_1 = require("../../../layout/src/LayoutSider");
const LayoutHeader_1 = require("../../../layout/src/LayoutHeader");
const utils_1 = require("./utils");
exports.default = vue_1.defineComponent({
    name: 'ServiceLayout',
    alias: ['NimbusServiceLayout'],
    props: {
        items: {
            type: Array,
            default: () => []
        },
        name: String,
        headerProps: Object,
        contentProps: Object,
        siderProps: Object,
        value: String,
        'onUpdate:value': Function,
        // deprecated
        paddingBody: {
            type: Boolean,
            default: false
        },
        onInput: Function,
        onSelect: Function,
        onExpandedNamesChange: Function
    },
    setup(props) {
        const router = vue_router_1.useRouter();
        const route = vue_router_1.useRoute();
        const bodyLayoutInstRef = vue_1.ref(null);
        const uncontrolledValueRef = vue_1.ref(null);
        const controlledValueRef = vue_1.toRef(props, 'value');
        const mergedValueRef = vooks_1.useMergedState(controlledValueRef, uncontrolledValueRef);
        const menuItemsRef = vue_1.computed(() => {
            return utils_1.createItems(props.items);
        });
        if (route) {
            syncPath(route.path);
            vue_1.watch(vue_1.toRef(route, 'path'), (path) => {
                syncPath(path);
            });
            vue_1.watch(vue_1.toRef(props, 'items'), () => {
                syncPath(route.path);
            });
        }
        function doUpdateValue(value) {
            const { onInput, onSelect, 'onUpdate:value': onUpdateValue } = props;
            uncontrolledValueRef.value = value;
            if (onSelect)
                onSelect(value);
            if (onInput)
                onInput(value);
            if (onUpdateValue)
                onUpdateValue(value);
        }
        function handleMenuUpdateValue(value, item) {
            if (router && item.path) {
                // path is copied to menu item in utils.createItems
                Promise.resolve(router.push(item.path)).catch((err) => {
                    console.log(err);
                });
            }
            doUpdateValue(value);
        }
        const scrollTo = (...args) => {
            var _a;
            ;
            ((_a = bodyLayoutInstRef.value) === null || _a === void 0 ? void 0 : _a.scrollTo)(...args);
        };
        function syncPath(path, items) {
            if (items === undefined)
                items = props.items;
            for (const item of items) {
                if (item.childItems || item.children) {
                    syncPath(path, item.childItems || item.children);
                }
                else if (item.path === path) {
                    doUpdateValue(item.name || item.key || '');
                    return;
                }
            }
        }
        return {
            bodyLayoutInstRef,
            scrollTo,
            handleMenuUpdateValue,
            menuItems: menuItemsRef,
            mergedValue: mergedValueRef,
            uncontrolledValue: uncontrolledValueRef
        };
    },
    render() {
        var _a, _b;
        const siderProps = Object.assign(Object.assign({ bordered: true }, this.siderProps), { showTrigger: true, collapseMode: 'width', showCollapsedContent: true, nativeScrollbar: false, collapsedWidth: 0, width: 288, triggerStyle: {
                top: 'calc(50% - 78px)'
            }, scrollbarProps: {
                style: {
                    width: '288px',
                    flexShrink: 0
                }
            }, style: (_a = this.siderProps) === null || _a === void 0 ? void 0 : _a.style });
        const contentProps = Object.assign(Object.assign({}, this.contentProps), { ref: 'bodyLayoutInstRef', nativeScrollbar: false, scrollbarProps: {
                // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
                contentStyle: {
                    width: '100%',
                    boxSizing: 'border-box',
                    padding: this.paddingBody ? '21px 48px' : ''
                }
            } });
        const headerProps = Object.assign(Object.assign({}, this.headerProps), { style: Object.assign(Object.assign({}, (_b = this.headerProps) === null || _b === void 0 ? void 0 : _b.style), { height: '64px' }), bordered: true });
        const navSlot = this.$slots.nav;
        return vue_1.h(Layout_1.default, {
            position: 'absolute'
        }, {
            default: () => [
                navSlot ? vue_1.h(LayoutHeader_1.default, headerProps, navSlot()) : null,
                vue_1.h(Layout_1.default, {
                    hasSider: true,
                    style: {
                        top: navSlot ? '64px' : undefined
                    },
                    position: 'absolute'
                }, {
                    default: () => [
                        vue_1.h(LayoutSider_1.default, siderProps, {
                            default: () => [
                                vue_1.h(SiderHeader_1.default, {
                                    name: this.name
                                }, {
                                    icon: this.$slots['drawer-header-icon']
                                }),
                                vue_1.h(menu_1.NMenu, {
                                    value: this.mergedValue,
                                    defaultExpandAll: true,
                                    rootIndent: 36,
                                    indent: 40,
                                    options: this.menuItems,
                                    'onUpdate:expandedKeys': this.onExpandedNamesChange,
                                    'onUpdate:value': this.handleMenuUpdateValue
                                })
                            ]
                        }),
                        vue_1.h(Layout_1.default, contentProps, {
                            default: this.$slots.default
                        })
                    ]
                })
            ]
        });
    }
});
