// naive-ui/lib/_deprecated/service-layout/src/SiderHeader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vue_1 = require("vue");
const icon_1 = require("../../../icon");
const divider_1 = require("../../../divider");
exports.default = vue_1.defineComponent({
    name: 'ServiceLayoutHeader',
    props: {
        name: String
    },
    render() {
        return this.name ? (vue_1.h(vue_1.Fragment, null,
            vue_1.h("div", { style: {
                    alignItems: 'center',
                    height: '64px',
                    paddingLeft: '36px',
                    fontSize: '16px',
                    fontWeight: 500,
                    display: 'flex',
                    position: 'relative'
                } },
                this.$slots.icon ? (vue_1.h(icon_1.NIcon, { size: 20, style: {
                        position: 'absolute',
                        left: '10px',
                        top: '50%',
                        transform: 'translateY(-50%)'
                    } }, { default: this.$slots.icon })) : null,
                vue_1.h("span", null, this.name)),
            vue_1.h(divider_1.NDivider, { style: {
                    margin: '0',
                    padding: '0 8px'
                } }))) : null;
    }
});
