// naive-ui/es/tree/src/TreeNodeContent.js
import { h, defineComponent, ref } from 'vue';
export default defineComponent({
    name: 'TreeNodeContent',
    props: {
        clsPrefix: {
            type: String,
            required: true
        },
        disabled: {
            type: Boolean,
            default: false
        },
        onClick: Function,
        onDragstart: Function
    },
    setup(props) {
        const selfRef = ref(null);
        function doClick(e) {
            const { onClick } = props;
            if (onClick)
                onClick(e);
        }
        function handleClick(e) {
            doClick(e);
        }
        return {
            selfRef,
            handleClick
        };
    },
    render() {
        const { clsPrefix, handleClick, onDragstart } = this;
        return (h("span", { ref: "selfRef", class: [`${clsPrefix}-tree-node-content`], onClick: handleClick, draggable: onDragstart === undefined ? undefined : true, onDragstart: onDragstart },
            h("div", { class: `${clsPrefix}-tree-node-content__text` }, this.$slots)));
    }
});
