// naive-ui/lib/_mixins/use-style.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vue_1 = require("vue");
const index_cssr_1 = require("../_styles/global/index.cssr");
const context_1 = require("../ssr/context");
const _utils_1 = require("../_utils");
function useStyle(mountId, style, clsPrefixRef) {
    if (!style) {
        if (process.env.NODE_ENV !== 'production')
            _utils_1.throwError('use-style', 'No style is specified.');
        return;
    }
    const ssrAdapter = vue_1.inject(context_1.ssrInjectionKey, undefined);
    const mountStyle = () => {
        const clsPrefix = clsPrefixRef === null || clsPrefixRef === void 0 ? void 0 : clsPrefixRef.value;
        style.mount({
            id: clsPrefix === undefined ? mountId : clsPrefix + mountId,
            head: true,
            props: {
                bPrefix: clsPrefix ? `.${clsPrefix}-` : undefined
            },
            ssr: ssrAdapter
        });
        index_cssr_1.default.mount({
            id: 'naive-ui/global',
            head: true,
            ssr: ssrAdapter
        });
    };
    if (ssrAdapter) {
        mountStyle();
    }
    else {
        vue_1.onBeforeMount(mountStyle);
    }
}
exports.default = useStyle;
