// naive-ui/es/popselect/src/Popselect.js
import { h, ref, provide, defineComponent } from 'vue';
import { NPopover } from '../../popover';
import { popoverBaseProps } from '../../popover/src/Popover';
import NPopselectPanel, { panelPropKeys, panelProps } from './PopselectPanel';
import { omit, keep } from '../../_utils';
import { useTheme } from '../../_mixins';
import { popselectLight } from '../styles';
import { popselectInjectionKey } from './interface';
const popselectProps = Object.assign(Object.assign(Object.assign(Object.assign({}, useTheme.props), popoverBaseProps), { 
    // eslint-disable-next-line vue/require-prop-types
    trigger: {
        type: String,
        default: 'hover'
    }, 
    // eslint-disable-next-line vue/require-prop-types
    showArrow: {
        type: Boolean,
        default: false
    } }), panelProps);
export default defineComponent({
    name: 'Popselect',
    props: popselectProps,
    setup(props) {
        const themeRef = useTheme('Popselect', 'Popselect', undefined, popselectLight, props);
        const popoverInstRef = ref(null);
        function syncPosition() {
            var _a;
            (_a = popoverInstRef.value) === null || _a === void 0 ? void 0 : _a.syncPosition();
        }
        function setShow(value) {
            var _a;
            (_a = popoverInstRef.value) === null || _a === void 0 ? void 0 : _a.setShow(value);
        }
        provide(popselectInjectionKey, {
            mergedThemeRef: themeRef,
            syncPosition,
            setShow
        });
        return {
            popoverInstRef,
            mergedTheme: themeRef
        };
    },
    render() {
        const { mergedTheme } = this;
        return h(NPopover, omit(this.$props, panelPropKeys, {
            padded: false,
            ref: 'popoverInstRef',
            internalExtraClass: 'popselect',
            theme: mergedTheme.peers.Popover,
            themeOverrides: mergedTheme.peerOverrides.Popover
        }), {
            trigger: this.$slots.default,
            default: () => {
                return h(NPopselectPanel, keep(this.$props, panelPropKeys));
            }
        });
    }
});
