// naive-ui/lib/dropdown/styles/light.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.self = void 0;
const seemly_1 = require("seemly");
const common_1 = require("../../_styles/common");
const use_theme_1 = require("../../_mixins/use-theme");
const _common_1 = require("./_common");
const styles_1 = require("../../popover/styles");
const self = (vars) => {
    const { primaryColor, textColor2, boxShadow2, dividerColor, hoverColor, popoverColor, invertedColor, borderRadius, fontSizeSmall, fontSizeMedium, fontSizeLarge, fontSizeHuge, heightSmall, heightMedium, heightLarge, heightHuge, textColor3 } = vars;
    return Object.assign(Object.assign({}, _common_1.default), { optionHeightSmall: heightSmall, optionHeightMedium: heightMedium, optionHeightLarge: heightLarge, optionHeightHuge: heightHuge, borderRadius, boxShadow: boxShadow2, fontSizeSmall,
        fontSizeMedium,
        fontSizeLarge,
        fontSizeHuge, 
        // non-inverted
        optionTextColor: textColor2, optionTextColorHover: textColor2, optionTextColorActive: primaryColor, optionTextColorChildActive: primaryColor, color: popoverColor, dividerColor, suffixColor: textColor2, prefixColor: textColor2, optionColorHover: hoverColor, optionColorActive: seemly_1.changeColor(primaryColor, { alpha: 0.1 }), groupHeaderTextColor: textColor3, 
        // inverted
        optionTextColorInverted: '#BBB', optionTextColorHoverInverted: '#FFF', optionTextColorActiveInverted: '#FFF', optionTextColorChildActiveInverted: '#FFF', colorInverted: invertedColor, dividerColorInverted: '#BBB', suffixColorInverted: '#BBB', prefixColorInverted: '#BBB', optionColorHoverInverted: primaryColor, optionColorActiveInverted: primaryColor, groupHeaderTextColorInverted: '#AAA' });
};
exports.self = self;
const dropdownLight = use_theme_1.createTheme({
    name: 'Dropdown',
    common: common_1.commonLight,
    peers: {
        Popover: styles_1.popoverLight
    },
    self: exports.self
});
exports.default = dropdownLight;
