// naive-ui/lib/config-consumer/src/ConfigConsumer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vue_1 = require("vue");
const ConfigProvider_1 = require("../../config-provider/src/ConfigProvider");
const use_legacy_1 = require("./use-legacy");
exports.default = vue_1.defineComponent({
    name: 'ConfigConsumer',
    props: {
        onNamespaceChange: {
            type: Function,
            default: undefined
        },
        /**
         * @deprecated
         */
        onThemeChange: {
            type: Function,
            default: undefined
        },
        /**
         * @deprecated
         */
        onLanguageChange: Function
    },
    setup(props) {
        const NConfigProvider = vue_1.inject(ConfigProvider_1.configProviderInjectionKey, null);
        if (NConfigProvider) {
            vue_1.watch(NConfigProvider.mergedNamespaceRef, (value, oldValue) => {
                const { onNamespaceChange } = props;
                if (onNamespaceChange)
                    onNamespaceChange(value, oldValue);
            });
        }
        return Object.assign({ NConfigProvider }, use_legacy_1.default(NConfigProvider, props));
    },
    render() {
        const defaultSlot = this.$slots.default;
        const { NConfigProvider } = this;
        return defaultSlot
            ? defaultSlot({
                namespace: NConfigProvider
                    ? NConfigProvider.mergedNamespaceRef.value
                    : null,
                // deprecated
                theme: this.legacyTheme,
                language: this.legacyLanguage,
                themeEnvironment: this.legacyThemeEnvironment,
                styleScheme: this.legacyStyleScheme
            })
            : [];
    }
});
